File: PHP/src/NyzoStringTransaction.php
<?php
  /**
   * ref: https://github.com/n-y-z-o/nyzoVerifier/blob/master/src/main/java/co/nyzo/verifier/nyzoString/NyzoStringTransaction.java
   * version: 0.0.2
   */
  require_once("NyzoString.php");
  require_once("NyzoStringType.php");
  require_once("utils/Buffer.php");
  require_once("utils/Int64.php");

  class NyzoStringTransaction implements NyzoString {

    private $timestamp;
    private $amount;
    private $receiverIdentifier;
    private $previousHashHeight;
    private $previousBlockHash;
    private $senderIdentifier;
    private $senderData;
    private $signature;

    public function __construct(Int64 $timestamp, Int64 $amount, Buffer $receiverIdentifier, Int64 $previousHashHeight, Buffer $previousBlockHash=null, Buffer $senderIdentifier, Buffer $senderData, Buffer $signature) {
      $senderData = $senderData->slice(0, 32);
      $this->timestamp = $timestamp;
      $this->amount = $amount;
      $this->receiverIdentifier = $receiverIdentifier;
      $this->previousHashHeight = $previousHashHeight;
      $this->previousBlockHash = $previousBlockHash;
      $this->senderIdentifier = $senderIdentifier;
      $this->senderData = $senderData;
      $this->signature = $signature;
    }

    public function getTimestamp(): Int64 {
      return $this->timestamp;
    }

    public function getAmount(): Int64 {
      return $this->amount;
    }

    public function getReceiverIdentifier(): Buffer {
      return $this->receiverIdentifier;
    }

    public function getPreviousHashHeight(): Int64 {
      return $this->previousHashHeight;
    }

    public function getPreviousBlockHash(): Buffer {
      return $this->previousBlockHash;
    }

    public function getSenderIdentifier(): Buffer {
      return $this->senderIdentifier;
    }

    public function getSenderData(): Buffer {
      return $this->senderData;
    }
    
    public function getSignature(): Buffer {
      return $this->signature;
    }

    /** @override */
    public function getType(): NyzoStringType {
      return NyzoStringType::Transaction();
    }

    /** @override */
    public function getBytes(): array {
      $senderDataLength = sizeof($this->senderData);
      $bytes = new Buffer(1 + 8 + 8 + 32 + 8 + 32 + 1 + $senderDataLength + 64);
      $bytes[0] = 2; // typeStandard
      $this->timestamp->copy($bytes, 1);
      $this->amount->copy($bytes, 1 + 8);
      $this->receiverIdentifier->copy($bytes, 1 + 8 + 8);
      $this->previousHashHeight->copy($bytes, 1 + 8 + 8 + 32);
      $this->senderIdentifier->copy($bytes, 1 + 8 + 8 + 32 + 8);
      $bytes[1 + 8 + 8 + 32 + 8 + 32] = $senderDataLength;
      $this->senderData->copy($bytes, 1 + 8 + 8 + 32 + 8 + 32 + 1);
      $this->signature->copy($bytes, 1 + 8 + 8 + 32 + 8 + 32 + 1 + $senderDataLength);
      return $bytes->toArray();
    }

    public static function fromByteBuffer(Buffer $buffer): NyzoStringTransaction {
      $index = 0;
      $type = $buffer->slice($index, $index + 1); // Ignored - only supports type 2, standard transaction
      $index += 1;
      $timestamp = new Int64($buffer->slice($index, $index + 8));
      $index += 8;
      $amount = new Int64($buffer->slice($index, $index + 8));
      $index += 8;
      $receiverBuffer = $buffer->slice($index, $index + 32);
      $index += 32;
      $previousHashHeight = new Int64($buffer->slice($index, $index + 8));
      $index += 8;
      $previousBlockHash = null;
      $senderBuffer = $buffer->slice($index, $index + 32);
      $index += 32;
      $senderDataLength = min($buffer[$index] & 0xff, 32);
      $index += 1;
      $dataBuffer = $buffer->slice($index, $index + $senderDataLength);
      $index += $senderDataLength;
      $signatureBuffer = $buffer->slice($index, $index + 64);

      return new NyzoStringTransaction($timestamp, $amount, $receiverBuffer, $previousHashHeight, $previousBlockHash, $senderBuffer, $dataBuffer, $signatureBuffer);
    }

    public static function fromHex(string $timestampHex, string $amountHex, string $receiverHex, string $previousHashHeightHex, string $previousBlockHashHex, string $senderHex, string $dataHex, string $signatureHex): NyzoStringTransaction {
      $timestamp = new Int64(Buffer::fromHex($timestampHex));
      $amount = new Int64(Buffer::fromHex($amountHex));
      $filteredString = substr(implode("", explode("-", $receiverHex)), 0, 64);
      $receiverBuffer = Buffer::fromHex($filteredString);
      $previousHashHeight = new Int64(Buffer::fromHex($previousHashHeightHex));
      $filteredHash = substr(implode("", explode("-", $previousBlockHashHex)), 0, 64);
      $previousBlockHashBuffer = Buffer::fromHex($filteredHash);
      $filteredSender = substr(implode("", explode("-", $senderHex)), 0, 64);
      $senderBuffer = Buffer::fromHex($filteredSender);
      $dataBuffer = Buffer::fromHex($dataHex);
      $signatureBuffer = Buffer::fromHex($signatureHex);
      return new NyzoStringTransaction($timestamp, $amount, $receiverBuffer, $previousHashHeight, $previousBlockHashBuffer, $senderBuffer, $dataBuffer, $signatureBuffer);
    }
  }

?>

File: Python/nyzostrings/nyzostringtransaction.py
"""
Nyzo String for transaction
"""

from nyzostrings.nyzostring import NyzoString


__version__ = "0.0.1"


class NyzoStringTransaction(NyzoString):
    def __init__(
        self,
        timestamp: int,
        amount: int,
        receiver_identifier: bytes,
        previous_hash_height: int,
        previous_block_hash: bytes,
        sender_identifier: bytes,
        sender_data: bytes,
        signature: bytes,
    ) -> None:
        index = 0
        sender_data = sender_data[:32]
        sender_data_len = len(sender_data)
        bytes_content = bytearray(1 + 8 + 8 + 32 + 8 + 32 + 1 + sender_data_len + 64)
        bytes_content[0] = 2  # Standard transaction
        index += 1
        bytes_content[index : index + 8] = timestamp.to_bytes(8, byteorder="big")
        index += 8
        bytes_content[index : index + 8] = amount.to_bytes(8, byteorder="big")
        index += 8
        bytes_content[index : index + 32] = receiver_identifier
        index += 32
        bytes_content[index : index + 8] = previous_hash_height.to_bytes(8, byteorder="big")
        index += 8
        bytes_content[index : index + 32] = sender_identifier
        index += 32
        bytes_content[index] = sender_data_len
        index += 1
        bytes_content[index : index + sender_data_len] = sender_data
        index += sender_data_len
        bytes_content[index : index + 64] = signature
        super().__init__("tx__", bytes_content)
        self.receiver_identifier = receiver_identifier
        self.sender_data = sender_data
        self.amount = amount
        self.timestamp = timestamp
        self.previous_hash_height = previous_hash_height
        self.previous_block_hash = previous_block_hash

    @classmethod
    def from_bytes(cls, byte_buffer: bytes) -> "NyzoStringTransaction":
        buffer = memoryview(byte_buffer)
        index = 0
        tx_type = buffer[index]  # Ignored - only supports type 2, standard transaction
        index += 1
        timestamp = int.from_bytes(
            buffer[index : index + 8], byteorder="big", signed=False
        )
        index += 8
        amount = int.from_bytes(
            buffer[index : index + 8], byteorder="big", signed=False
        )
        index += 8
        receiver_buffer = buffer[index : index + 32]
        index += 32
        previous_hash_height = int.from_bytes(
            buffer[index : index + 8], byteorder="big", signed=False
        )
        index += 8
        previous_block_hash = None
        sender_buffer = buffer[index : index + 32]
        index += 32
        sender_data_length = min(buffer[index] & 0xFF, 32)
        index += 1
        data_buffer = buffer[index : index + sender_data_length]
        index += sender_data_length
        signature_buffer = buffer[index : index + 64]
        return NyzoStringTransaction(
            timestamp,
            amount,
            receiver_buffer,
            previous_hash_height,
            previous_block_hash,
            sender_buffer,
            data_buffer,
            signature_buffer,
        )

    @classmethod
    def from_hex(
        cls,
        timestamp_hex: str,
        amount_hex: str,
        receiver_identifier_hex: str,
        previous_hash_height_hex: str,
        previous_block_hash_hex: str,
        sender_identifier_hex: str,
        data_hex: str,
        signature_hex: str,
    ) -> "NyzoStringTransaction":
        timestamp = int(timestamp_hex, 16)
        amount = int(amount_hex, 16)
        filtered_string = receiver_identifier_hex.replace("-", "")[:64]
        receiver_identifier = bytes.fromhex(filtered_string)
        previous_hash_height = int(previous_hash_height_hex, 16)
        filtered_string = previous_block_hash_hex.replace("-", "")[:64]
        previous_block_hash = bytes.fromhex(filtered_string)
        filtered_string = sender_identifier_hex.replace("-", "")[:64]
        sender_identifier = bytes.fromhex(filtered_string)
        filtered_string = data_hex.replace("-", "")[:64]
        sender_data = bytes.fromhex(filtered_string)
        filtered_string = signature_hex.replace("-", "")[:128]
        signature = bytes.fromhex(filtered_string)
        return NyzoStringTransaction(timestamp, amount, receiver_identifier, previous_hash_height, previous_block_hash, sender_identifier, sender_data, signature
        )
